pkg/storage/redis/redis.go

```diff
@@ -1,404 +1,393 @@
 	"time"
 
-	"github.com/cenkalti/backoff/v4"
+	redis "github.com/go-redis/redis/v8"
 	"github.com/golang/protobuf/proto"
-	"github.com/golang/protobuf/ptypes"
-	"github.com/gomodule/redigo/redis"
 	"google.golang.org/protobuf/types/known/anypb"
+	"google.golang.org/protobuf/types/known/timestamppb"
 
 	"github.com/pomerium/pomerium/config"
 	"github.com/pomerium/pomerium/internal/log"
+	"github.com/pomerium/pomerium/internal/signal"
 	"github.com/pomerium/pomerium/internal/telemetry/metrics"
 	"github.com/pomerium/pomerium/internal/telemetry/trace"
 	"github.com/pomerium/pomerium/pkg/grpc/databroker"
-	"github.com/pomerium/pomerium/pkg/storage"
 )
 
-// Name is the storage type name for redis backend.
+// Name of the storage backend.
 const Name = config.StorageRedisName
 
-var _ storage.Backend = (*DB)(nil)
+const (
+	maxTransactionRetries = 100
+	watchPollInterval     = 30 * time.Second
+)
+
+// custom errors
+var (
+	ErrExceededMaxRetries = errors.New("redis: transaction reached maximum number of retries")
+)
 
-// DB wraps redis conn to interact with redis server.
+// DB implements the storage.Backend on top of redis.
 type DB struct {
-	pool                  *redis.Pool
-	recordType            string
-	lastVersionKey        string
-	lastVersionChannelKey string
-	versionSet            string
-	deletedSet            string
-	tlsConfig             *tls.Config
-	notifyChMu            sync.Mutex
+	cfg *dbConfig
+
+	client *redis.Client
 
 	closeOnce sync.Once
 	closed    chan struct{}
 }
 
-// New returns new DB instance.
-func New(rawURL, recordType string, opts ...Option) (*DB, error) {
+// New creates a new redis storage backend.
+func New(rawURL string, options ...Option) (*DB, error) {
 	db := &DB{
-		recordType:            recordType,
-		versionSet:            recordType + "_version_set",
-		deletedSet:            recordType + "_deleted_set",
-		lastVersionKey:        recordType + "_last_version",
-		lastVersionChannelKey: recordType + "_last_version_ch",
-		closed:                make(chan struct{}),
+		cfg:    getConfig(options...),
+		closed: make(chan struct{}),
 	}
-
-	for _, o := range opts {
-		o(db)
+	opts, err := redis.ParseURL(rawURL)
+	if err != nil {
+		return nil, err
 	}
-	db.pool = &redis.Pool{
-		Wait: true,
-		Dial: func() (redis.Conn, error) {
-			c, err := redis.DialURL(rawURL, redis.DialTLSConfig(db.tlsConfig))
-			if err != nil {
-				return nil, fmt.Errorf(`redis.DialURL(): %w`, err)
-			}
-			return c, nil
-		},
-		TestOnBorrow: func(c redis.Conn, t time.Time) error {
-			if time.Since(t) < time.Minute {
-				return nil
-			}
-			_, err := c.Do("PING")
-			if err != nil {
-				return fmt.Errorf(`c.Do("PING"): %w`, err)
-			}
-			return nil
-		},
-		MaxIdle:     64,
-		IdleTimeout: time.Minute,
-		MaxActive:   128,
+	// when using TLS, the TLS config will not be set to nil, in which case we replace it with our own
+	if opts.TLSConfig != nil {
+		opts.TLSConfig = db.cfg.tls
 	}
-	metrics.AddRedisMetrics(db.pool.Stats)
```

Comment:
This metrics instrumentation is redis client specific.  Can fix in a follow for compatibility with https://pkg.go.dev/github.com/go-redis/redis/v8@v8.4.2/internal/pool#Stats